pkg/connection/connection.go

```diff
@@ -0,0 +1,337 @@
+		}
+		glog.V(4).Infof("Still trying, connection is %s", conn.GetState())
+	}
+}
+
+func (c *csiConnection) GetDriverName(ctx context.Context) (string, error) {
+	client := csi.NewIdentityClient(c.conn)
+
+	req := csi.GetPluginInfoRequest{}
+
+	rsp, err := client.GetPluginInfo(ctx, &req)
+	if err != nil {
+		return "", err
+	}
+	name := rsp.GetName()
+	if name == "" {
+		return "", fmt.Errorf("name is empty")
+	}
+	return name, nil
+}
+
+func (c *csiConnection) Probe(ctx context.Context) error {
+	client := csi.NewIdentityClient(c.conn)
+
+	req := csi.ProbeRequest{}
+
+	_, err := client.Probe(ctx, &req)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *csiConnection) SupportsControllerCreateSnapshot(ctx context.Context) (bool, error) {
+	client := csi.NewControllerClient(c.conn)
+	req := csi.ControllerGetCapabilitiesRequest{}
+
+	rsp, err := client.ControllerGetCapabilities(ctx, &req)
+	if err != nil {
+		return false, err
+	}
+	caps := rsp.GetCapabilities()
+	for _, cap := range caps {
+		if cap == nil {
+			continue
+		}
+		rpc := cap.GetRpc()
+		if rpc == nil {
+			continue
+		}
+		if rpc.GetType() == csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (c *csiConnection) SupportsControllerListSnapshots(ctx context.Context) (bool, error) {
+	client := csi.NewControllerClient(c.conn)
+	req := csi.ControllerGetCapabilitiesRequest{}
+
+	rsp, err := client.ControllerGetCapabilities(ctx, &req)
+	if err != nil {
+		return false, err
+	}
+	caps := rsp.GetCapabilities()
+	for _, cap := range caps {
+		if cap == nil {
+			continue
+		}
+		rpc := cap.GetRpc()
+		if rpc == nil {
+			continue
+		}
+		if rpc.GetType() == csi.ControllerServiceCapability_RPC_LIST_SNAPSHOTS {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (c *csiConnection) CreateSnapshot(ctx context.Context, snapshot *crdv1.VolumeSnapshot, volume *v1.PersistentVolume, parameters map[string]string) (*crdv1.VolumeSnapshotData, *crdv1.VolumeSnapshotStatus, error) {
+	glog.V(5).Infof("CSI CreateSnapshot: %s", snapshot.Name)
+	if volume.Spec.CSI == nil {
+		return nil, nil, fmt.Errorf("CSIPersistentVolumeSource not defined in spec")
+	}
+
+	client := csi.NewControllerClient(c.conn)
+
+	driverName, err := c.GetDriverName(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	req := csi.CreateSnapshotRequest{
+		SourceVolumeId:        volume.Spec.CSI.VolumeHandle,
+		Name:                  snapshot.Name,
+		Parameters:            parameters,
+		CreateSnapshotSecrets: nil,
+	}
+
+	rsp, err := client.CreateSnapshot(ctx, &req)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	snapDataName := GetSnapshotDataNameForSnapshot(snapshot)
+
+	persistentVolumeRef, err := ref.GetReference(scheme.Scheme, volume)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unexpected error getting volume reference: %v", err)
+	}
+
+	// Create VolumeSnapshotData in the database
+	snapshotData := &crdv1.VolumeSnapshotData{
```

Comment:
I feel it is better to create VolumeSnapshotData object in the controller, not in connection code.

Reply:
Sure.